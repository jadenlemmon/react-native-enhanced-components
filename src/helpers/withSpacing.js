import React from 'react';
import { pick, get } from 'lodash';
import spacing, { KEYS } from '../styles/spacing';

export default function withSpacing(Component) {
    return class extends React.Component {
        get styles() {
            const { props } = this;

            // get style prop styles
            const styleProp = get(props, 'style', {});

            // get spacing props and map to proper values
            const grabbedStyles = Object.keys(pick(props, KEYS)).map(
                key => spacing[`${key}${props[key]}`]
            );

            // grab other styles
            const other = Array.isArray(styleProp) ? styleProp : [styleProp];

            // merge all
            return [...grabbedStyles, ...other];
        }

        render() {
            return <Component {...this.props} style={this.styles} />;
        }
    };
}
