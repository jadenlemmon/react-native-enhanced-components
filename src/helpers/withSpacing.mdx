---
name: With Spacing HOC
---

import { Playground, PropsTable } from 'docz';
import { TouchableOpacity, Text } from 'react-native';
import withSpacing from './withSpacing';

# With Spacing HOC

## Minimel Example

The With Styles HOC provides a spacing wrapper that can wrap any component and provide spacing utilility props.

<Playground>
  {() => {
    const MyCustomButton = withSpacing(TouchableOpacity);
    return (
        <MyCustomButton p={50} style={{ backgroundColor: 'blue' }}>
          <Text style={{ color: 'white', textAlign: 'center' }}>My Button</Text>
        </MyCustomButton>
    );
}}

</Playground>
