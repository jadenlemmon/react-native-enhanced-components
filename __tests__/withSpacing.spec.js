import React from 'react';
import renderer from 'react-test-renderer';
import { View } from 'react-native';
import withSpacing from '../src/helpers/withSpacing';

const FooBar = withSpacing(View);

test('it should render component with backgroundColor', () => {
  const tree = renderer.create(<FooBar mt={10} />).toJSON();
  expect(tree.props.style[0]).toHaveProperty('marginTop');
  expect(tree).toMatchSnapshot();
});
