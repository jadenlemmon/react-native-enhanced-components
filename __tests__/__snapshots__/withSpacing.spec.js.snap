// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render component with backgroundColor 1`] = `
<View
  mt={10}
  style={
    Array [
      Object {
        "marginTop": 10,
      },
      Object {},
    ]
  }
/>
`;
